# Analysis of Successful Movies (Notebook 1)
* Benjamin Grossmann

This notebook retrieves the data from online.
Then it performs some preprocessing and filtering to keep only the movies that meet the desired criteria.
The final steps are to save the reduced data set.

After the reduced data set has been saved, further work on this project should use Notebook 2. This will reduce the time to bring the data into a project-ready state.

If the reduced data set should need to be reset to its initial condition, then re-run Notebook 1.

THe IMDB files being used:
* https://datasets.imdbws.com/title.basics.tsv.gz
* https://datasets.imdbws.com/title.akas.tsv.gz
* https://datasets.imdbws.com/title.ratings.tsv.gz

The data dictionay:
* https://www.imdb.com/interfaces/

## Load Data

In [1]:
import numpy as np
import pandas as pd
from sqlalchemy import create_engine

In [3]:
%%time
# This code block can take around 5 minutes

# Store the data locations as strings
url_basics = 'https://datasets.imdbws.com/title.basics.tsv.gz'
url_akas = 'https://datasets.imdbws.com/title.akas.tsv.gz'
url_ratings = 'https://datasets.imdbws.com/title.ratings.tsv.gz'

# Retrieve the data from online
basics = pd.read_csv(url_basics, sep='\t', low_memory=False)
akas = pd.read_csv(url_akas, sep='\t', low_memory=False)
ratings = pd.read_csv(url_ratings, sep='\t', low_memory=False)

Wall time: 4min 20s


In [58]:
# Display basic information about the data
print(f"The basics dataframe shape: {basics.shape}")
print(f"The akas dataframe shape: {akas.shape}")
print(f"The ratings dataframe shape: {ratings.shape}")

The basics dataframe shape: (81040, 9)
The akas dataframe shape: (1322196, 8)
The ratings dataframe shape: (1245178, 3)


## Standardizing Missing Values

In [5]:
%%time
# This code block can take around 1 minutes

# Replace the \N placeholders with np.nan
basics = basics.replace({'\\N':np.nan})
akas = akas.replace({'\\N':np.nan})
ratings = ratings.replace({'\\N':np.nan})

Wall time: 52.3 s


## Filter the Akas DataFrame

In [38]:
akas.info()
akas.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32014153 entries, 0 to 32014152
Data columns (total 8 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   titleId          object
 1   ordering         int64 
 2   title            object
 3   region           object
 4   language         object
 5   types            object
 6   attributes       object
 7   isOriginalTitle  object
dtypes: int64(1), object(7)
memory usage: 1.9+ GB


titleId  ordering                      title region language  \
0  tt0000001         1                 Карменсіта     UA      NaN   
1  tt0000001         2                 Carmencita     DE      NaN   
2  tt0000001         3  Carmencita - spanyol tánc     HU      NaN   
3  tt0000001         4                 Καρμενσίτα     GR      NaN   
4  tt0000001         5                 Карменсита     RU      NaN   

         types     attributes isOriginalTitle  
0  imdbDisplay            NaN               0  
1          NaN  literal title               0  
2  imdbDisplay            NaN               0  
3  imdbDisplay            NaN               0  
4  imdbDisplay            NaN               0

First, the 'region' column will be filtered to contain only the value 'US'.

In [59]:
print(f'The akas dataframe shape: {akas.shape}')

print(f'Eliminate entries that are not release in the US')
is_US = (akas['region']=='US')
akas = akas.loc[ is_US , :]

print(f'The akas dataframe shape: {akas.shape}')
akas.head()

The akas dataframe shape: (1322196, 8)
Eliminate entries that are not release in the US
The akas dataframe shape: (1322196, 8)


titleId  ordering                   title region language        types  \
5   tt0000001         6              Carmencita     US      NaN  imdbDisplay   
14  tt0000002         7  The Clown and His Dogs     US      NaN          NaN   
33  tt0000005        10        Blacksmith Scene     US      NaN  imdbDisplay   
36  tt0000005         1     Blacksmithing Scene     US      NaN  alternative   
41  tt0000005         6     Blacksmith Scene #1     US      NaN  alternative   

               attributes isOriginalTitle  
5                     NaN               0  
14  literal English title               0  
33                    NaN               0  
36                    NaN               0  
41                    NaN               0

The 'isOriginalTitle' column is object data type, but appears like it should be integer data type.

In [51]:
akas['isOriginalTitle'].unique()

array(['0', nan, '1'], dtype=object)

So, 'isOriginalTitle' does contain string numerical values, that should be converted. Since nan is considered a float, I will convert the string numbers to floats.

In [61]:
akas['isOriginalTitle'] = akas['isOriginalTitle'].apply(float)


5           0.0
14          0.0
33          0.0
36          0.0
41          0.0
           ... 
32013824    0.0
32013862    0.0
32013864    0.0
32013881    0.0
32013897    0.0
Name: isOriginalTitle, Length: 1322196, dtype: float64

In [47]:
akas.head()

titleId  ordering                   title region language        types  \
5   tt0000001         6              Carmencita     US      NaN  imdbDisplay   
14  tt0000002         7  The Clown and His Dogs     US      NaN          NaN   
33  tt0000005        10        Blacksmith Scene     US      NaN  imdbDisplay   
36  tt0000005         1     Blacksmithing Scene     US      NaN  alternative   
41  tt0000005         6     Blacksmith Scene #1     US      NaN  alternative   

               attributes isOriginalTitle  
5                     NaN               0  
14  literal English title               0  
33                    NaN               0  
36                    NaN               0  
41                    NaN               0

## Filter/Clean the Basics Dataframe

### Drop Missing Genres Or Runtimes

In [8]:
# Removing movies with missing values for genre or runtime

print(f'The basics dataframe shape: {basics.shape}')

has_runtimeMinutes = ~basics['runtimeMinutes'].isna()
has_genres = ~basics['genres'].isna()

print(f'Eliminate entries that are missing runtime or genre')
basics = basics.loc[ (has_runtimeMinutes) & (has_genres) , : ]

print(f'The basics dataframe shape: {basics.shape}')

The basics dataframe shape: (8928653, 9)
Eliminate entries that are missing runtime or genre
The basics dataframe shape: (2336224, 9)


In [11]:
basics['runtimeMinutes'] = basics['runtimeMinutes'].apply(int)
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2336224 entries, 0 to 8928652
Data columns (total 9 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   tconst          object
 1   titleType       object
 2   primaryTitle    object
 3   originalTitle   object
 4   isAdult         object
 5   startYear       object
 6   endYear         object
 7   runtimeMinutes  int64 
 8   genres          object
dtypes: int64(1), object(8)
memory usage: 178.2+ MB


### Keep Full-Length Movies

In [14]:
basics['titleType'].unique().tolist()

['short',
 'movie',
 'tvEpisode',
 'tvSeries',
 'tvShort',
 'tvMovie',
 'tvMiniSeries',
 'video',
 'tvSpecial',
 'videoGame']

In [15]:
# Removing movies without a string value
print(f'The basics dataframe shape: {basics.shape}')

print(f'Eliminate entries that are not full-length movies')
is_movie = (basics['titleType'] == 'movie')
basics = basics.loc[ is_movie , : ]

print(f'The basics dataframe shape: {basics.shape}')

The basics dataframe shape: (2336224, 9)
Eliminate entries that are not full-length movies
The basics dataframe shape: (360501, 9)


In [16]:
basics['titleType'].unique().tolist()

['movie']

### Drop Documentaries

In [17]:
# Find all the genres present in the basics dataframe
# Because each entry can belong to multiple genres,
# some extra steps are needed to show each unique genre.

def splitter(n):
    return n.split(",")

set(list(np.concatenate(list(map( splitter , basics['genres'].unique().tolist() ))).flat))

{'Action',
 'Adult',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Family',
 'Fantasy',
 'Film-Noir',
 'Game-Show',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'News',
 'Reality-TV',
 'Romance',
 'Sci-Fi',
 'Short',
 'Sport',
 'Talk-Show',
 'Thriller',
 'War',
 'Western'}

In [18]:
# Removing rows with a sub-string value
print(f'The basics dataframe shape: {basics.shape}')
print(f'Eliminate entries that are documentaries')

is_documentary = basics['genres'].str.contains('Documentary', case=False)
# note: case parameter means case-sensitive spelling

basics = basics.loc[ ~is_documentary , :]

print(f'The basics dataframe shape: {basics.shape}')

The basics dataframe shape: (360501, 9)
Eliminate entries that are documentaries
The basics dataframe shape: (274436, 9)


In [19]:
# Find all the genres present in the basics dataframe
# Because each entry can belong to multiple genres,
# some extra steps are needed to show each unique genre.

def splitter(n):
    return n.split(",")

set(list(np.concatenate(list(map( splitter , basics['genres'].unique().tolist() ))).flat))

{'Action',
 'Adult',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Drama',
 'Family',
 'Fantasy',
 'Film-Noir',
 'Game-Show',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'News',
 'Reality-TV',
 'Romance',
 'Sci-Fi',
 'Short',
 'Sport',
 'Talk-Show',
 'Thriller',
 'War',
 'Western'}

### Keep Released 21st Century Movies

The 'startYear' column is of the object data type, so the values should be checked. Ideally, the years are integers. Because np.nan values are floats, I will convert any string numbers to floats to avoid an error in trying to convert np.nan values to integers.

In [21]:
startYear_null_count = basics['startYear'].isna().sum()
startYear_null_percent = 100*startYear_null_count/basics.shape[0]
startYear_min = basics['startYear'].apply(float).min()
startYear_max = basics['startYear'].apply(float).max()

print(f"The basics dataframe has {startYear_null_count} null startYear values.")
print(f"\nThe basics dataframe:")
print(basics['startYear'].apply(type).value_counts())
print(f"\nNull values constitute {startYear_null_percent:.2f}% of the basics dataframe")
print(f"The range of years is {startYear_min} through {startYear_max}")

The basics dataframe has 4155 null startYear values.

The basics dataframe:
<class 'str'>      270281
<class 'float'>      4155
Name: startYear, dtype: int64

Null values constitute 1.51% of the basics dataframe
The range of years is 1903.0 through 2027.0


I will drop the null values. They constitute only 1.51% of the movies. Imputing any values could potentially be extremely misleading, as characteristics of successful movies have changed in the past century.

By filtering the 'startYear' by inequalities, the np.nan values will be filtered out as they not less than or greater then a numerical value.

In [31]:
print(f'The basics dataframe shape: {basics.shape}')
print(f'Eliminate entries that before the year 2000')
print(f'Eliminate entries that are after the year 2022')

is_released = (basics['startYear'].apply(float) <= 2022)
is_year_20XX = (basics['startYear'].apply(float) >= 2000)

basics = basics.loc[ is_released & is_year_20XX, : ]

print(f'The basics dataframe shape: {basics.shape}')

The basics dataframe shape: (274436, 9)
Eliminate entries that before the year 2000
Eliminate entries that are after the year 2022
The basics dataframe shape: (139319, 9)


In [34]:
basics['startYear'] = basics['startYear'].apply(int)
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 139319 entries, 34805 to 8928502
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          139319 non-null  object
 1   titleType       139319 non-null  object
 2   primaryTitle    139319 non-null  object
 3   originalTitle   139319 non-null  object
 4   isAdult         139319 non-null  object
 5   startYear       139319 non-null  int64 
 6   endYear         0 non-null       object
 7   runtimeMinutes  139319 non-null  int64 
 8   genres          139319 non-null  object
dtypes: int64(2), object(7)
memory usage: 10.6+ MB


0

## 

In [48]:
print(f'The basics dataframe shape: {basics.shape}')

# Filter out entries from the basics dataframe that are not released in the US
is_basic_in_US =  basics['tconst'].isin(akas['titleId'])
basics = basics.loc[ is_basic_in_US , :]

print(f'The basics dataframe shape: {basics.shape}')
basics.head()

The basics dataframe shape: (139319, 9)
The basics dataframe shape: (81040, 9)


tconst titleType                                       primaryTitle  \
34805  tt0035423     movie                                     Kate & Leopold   
61119  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
67672  tt0069049     movie                         The Other Side of the Wind   
86806  tt0088751     movie                                  The Naked Monster   
91077  tt0093119     movie                                Grizzly II: Revenge   

                                   originalTitle isAdult  startYear endYear  \
34805                             Kate & Leopold       0       2001     NaN   
61119  El Tango del Viudo y Su Espejo Deformante       0       2020     NaN   
67672                 The Other Side of the Wind       0       2018     NaN   
86806                          The Naked Monster       0       2005     NaN   
91077                   Grizzly II: The Predator       0       2020     NaN   

       runtimeMinutes                  genres  
34805             118  Comedy,Fantasy,Romance  
61119              70                   Drama  
67672             122                   Drama  
86806             100    Comedy,Horror,Sci-Fi  
91077              74   Horror,Music,Thriller

# XYZ

In [ ]:
# This code was was used to create a Data folder for saving downloaded data


# # make new folder with os
# import os
# os.makedirs('Data', exist_ok = True)
# # verify folder was created
# os.listdir('Data/')

In [ ]:
# save dataframes to file
basics.to_csv('Data/title_basics.csv.gz', compression='gzip', index=False)
akas.to_csv('Data/title_akas.csv.gz', compression='gzip', index=False)
ratings.to_csv('Data/title_ratings.csv.gz', compression='gzip', index=False)

In [ ]:
# open saved files and preview
basics = pd.read_csv('Data/title_basics.csv.gz', low_memory = False)
akas = pd.read_csv('Data/title_akas.csv.gz', low_memory = False)
ratings = pd.read_csv('Data/title_ratings.csv.gz', low_memory = False)

In [ ]:
display(basics.head())
display(akas.head())
display(ratings.head())